ArmPkg/Drivers/SmmuDxe/SmmuDxe.c

```diff
@@ -0,0 +1,774 @@
+  // Set the maximum output address width. Per SMMUv3.2 spec (sections 5.2 and
+  // 3.4.1), the maximum input address width with AArch64 format is given by
+  // SMMU_IDR5.OAS field and capped at:
+  // - 48 bits in SMMUv3.0,
+  // - 52 bits in SMMUv3.1+. However, an address greater than 48 bits can
+  //   only be output from stage 2 when a 64KB translation granule is in use
+  //   for that translation table, which is not currently supported (only 4KB
+  //   granules).
+  //
+  //  Thus the maximum input address width is restricted to 48-bits even if
+  //  it is advertised to be larger.
+  //
+  OutputAddressWidth = SmmuV3DecodeAddressWidth (Idr5.Oas);
+
+  if (OutputAddressWidth < 48) {
+    StreamEntry->S2Ps = SmmuV3EncodeAddressWidth (OutputAddressWidth);
+  } else {
+    StreamEntry->S2Ps = SmmuV3EncodeAddressWidth (48);
+  }
+
+  InputSize           = OutputAddressWidth;
+  StreamEntry->S2T0Sz = 64 - InputSize;
+  if (IortCohac != 0) {
+    StreamEntry->S2Ir0 = ARM64_RGNCACHEATTR_WRITEBACK_WRITEALLOCATE;
+    StreamEntry->S2Or0 = ARM64_RGNCACHEATTR_WRITEBACK_WRITEALLOCATE;
+    StreamEntry->S2Sh0 = ARM64_SHATTR_INNER_SHAREABLE;
+  } else {
+    StreamEntry->S2Ir0 = ARM64_RGNCACHEATTR_NONCACHEABLE;
+    StreamEntry->S2Or0 = ARM64_RGNCACHEATTR_NONCACHEABLE;
+    StreamEntry->S2Sh0 = ARM64_SHATTR_OUTER_SHAREABLE;
+  }
+
+  StreamEntry->S2Rs = 0x2;   // record faults
+
+  if (Idr1.AttrTypesOvr != 0) {
+    StreamEntry->ShCfg = 0x1;
+  }
+
+  if ((Idr1.AttrTypesOvr != 0) && ((CCA == 1) && (CPM == 1) && (DACS == 0))) {
+    StreamEntry->Mtcfg   = 0x1;
+    StreamEntry->MemAttr = 0xF;   // Inner+Outer write-back cached
+    StreamEntry->ShCfg   = 0x3;   // Inner shareable
+  }
+
+  StreamEntry->Valid = 1;
+
+  Status = EFI_SUCCESS;
+  return Status;
+}
+
+/**
+  Allocate a stream table for SMMUv3.
+
+  @param [in]  SmmuInfo       Pointer to the SMMU_INFO structure.
+  @param [in]  SmmuConfig     Pointer to the SMMU configuration.
+  @param [out] Log2Size       Pointer to store the log2 size of the stream table.
+  @param [out] Size           Pointer to store the size of the stream table.
+
+  @retval Pointer to the allocated stream table.
+**/
+SMMUV3_STREAM_TABLE_ENTRY *
+EFIAPI
+SmmuV3AllocateStreamTable (
+  IN SMMU_INFO    *SmmuInfo,
+  IN SMMU_CONFIG  *SmmuConfig,
+  OUT UINT32      *Log2Size,
+  OUT UINT32      *Size
+  )
+{
+  UINT32  MaxStreamId;
+  UINT32  SidMsb;
+  UINT32  Alignment;
+  UINTN   Pages;
+  VOID    *AllocatedAddress;
+
+  MaxStreamId      = SmmuConfig->Config.SmmuNode.SmmuIdMap.OutputBase + SmmuConfig->Config.SmmuNode.SmmuIdMap.NumIds;
+  SidMsb           = HighBitSet32 (MaxStreamId);
+  *Log2Size        = SidMsb + 1;
+  *Size            = SMMUV3_LINEAR_STREAM_TABLE_SIZE_FROM_LOG2 (*Log2Size);
+  *Size            = ROUND_UP (*Size, EFI_PAGE_SIZE);
+  Alignment        = ALIGN_UP_BY (*Size, EFI_PAGE_SIZE);
+  Pages            = EFI_SIZE_TO_PAGES (*Size);
+  AllocatedAddress = AllocateAlignedPages (Pages, Alignment);

```

Comment:
Does this actually need to be aligned on the number of bytes required? i.e. if you needed 0x2048 bytes for the table, this call would ensure that `AllocatedAddress` is on a multiple of 0x2048. I'd assume you really just need it to be page aligned? In which case same comment on using `AllocatePages`.

Reply:
The smmu spec requires linear page tables to be aligned to their size.  Adding comments to explain.